### PR TITLE
Add spread to list/record matching

### PR DESCRIPTION
**Pull request by gregorybchris**

### CODE REVIEW

File: scrapscript.py [hunk 1]
```diff
@@ -280,26 +273,36 @@ def parse(tokens: typing.List[str], p: float = 0) -> "Object":
...
+        while True:
+            # TODO: Implement .. operator
+            if tokens[0] == "}":
+                tokens.pop(0)
+                break
+            elif tokens[0] == ",":
+                tokens.pop(0)
+            elif tokens[0] == "...":
+                # TODO(chris): Revisit how spread is saved to record
+                l.data["__spread"] = Spread()
```
[the remark this answers is not in the document]
Comment (gregorybchris) [reply scrapscript.py]: Updated to ...

File: scrapscript.py [hunk 2]
```diff
@@ -621,30 +629,41 @@ def match(obj: Object, pattern: Object) -> Optional[Env]:
...
             return None
         result: Env = {}
-        for key, value in pattern.data.items():
-            obj_value = obj.data.get(key)
-            if obj_value is None:
+        use_spread = False
+        for key, pattern_item in pattern.data.items():
+            if isinstance(pattern_item, Spread):
+                use_spread = True
+                continue
```
[the remark this answers is not in the document]
Comment (gregorybchris) [reply scrapscript.py]: done broke

File: scrapscript.py [hunk 3]
```diff
@@ -621,30 +629,41 @@ def match(obj: Object, pattern: Object) -> Optional[Env]:
...
             return None
-        if len(pattern.items) != len(obj.items):
-            # TODO: Remove this check when implementing ... operator
-            return None
         result: Env = {}  # type: ignore
+        use_spread = False
         for i, pattern_item in enumerate(pattern.items):
+            if isinstance(pattern_item, Spread):
+                use_spread = True
+                continue
```
[the remark this answers is not in the document]
Comment (gregorybchris) [reply scrapscript.py]: brrrrrkkk

File: scrapscript.py [hunk 4]
```diff
@@ -280,26 +273,36 @@ def parse(tokens: typing.List[str], p: float = 0) -> "Object":
...
-            while tokens.pop(0) != "]":
-                # TODO: Implement .. and ... operators
+        while True:
+            # TODO: Implement .. operator
+            if tokens[0] == "]":
+                tokens.pop(0)
+                break
+            elif tokens[0] == ",":
+                tokens.pop(0)
+            elif tokens[0] == "...":
```
[the remark this answers is not in the document]
Comment (gregorybchris) [reply scrapscript.py]: Do you remember what this was referring to?

[the remark this answers is not in the document]
Comment (tekknolagi) [reply scrapscript.py]: I think see my branch that built on top of this -- it didn't change the list case so much

[the remark this answers is not in the document]
Comment (tekknolagi) [reply scrapscript.py]: https://github.com/tekknolagi/scrapscript/blob/mb-spread-mods/scrapscript.py#L276-L286

**Comment by tekknolagi** [comment]
Try rebasing on my tokenizer fix

File: scrapscript.py [hunk 5]
```diff
@@ -299,6 +299,7 @@ def xp(n: float) -> Prec:
     "!": lp(3),
     ".": rp(3),
     "?": rp(3),
+    "...": xp(2),
     ",": xp(1),
     # TODO: Fix precedence for spread
     "...": xp(0),
```
Comment (tekknolagi) [inline scrapscript.py]: delet

File: scrapscript.py [hunk 6]
```diff
@@ -374,26 +367,36 @@ def parse(tokens: typing.List[Token], p: float = 0) -> "Object":
...
-        token = tokens[0]
-        if isinstance(token, RightBracket):
-            tokens.pop(0)
-        else:
-            l.items.append(parse(tokens, 2))
-            while not isinstance(tokens.pop(0), RightBracket):
-                # TODO: Implement .. and ... operators
+        while True:
+            # TODO: Implement .. operator
+            if isinstance(tokens[0], RightBracket):
```
Comment (tekknolagi) [inline scrapscript.py]: I think you should keep the older structure (see my branch). As it is, your code seems to allow `[,,,,,,]`

File: scrapscript.py [hunk 7]
```diff
@@ -374,26 +367,36 @@ def parse(tokens: typing.List[Token], p: float = 0) -> "Object":
...
-            while not isinstance(tokens.pop(0), RightBrace):
-                # TODO: Implement .. and ... operators
-                assign = parse_assign(tokens, 2)
-                l.data[assign.name.name] = assign.value
+        while True:
+            # TODO: Implement .. operator
+            if isinstance(tokens[0], RightBrace):
+                tokens.pop(0)
+                break
+            elif tokens[0] == Operator(","):
```
Comment (tekknolagi) [inline scrapscript.py]: As above

File: scrapscript.py [hunk 8]
```diff
@@ -384,6 +384,8 @@ def parse(tokens: typing.List[Token], p: float = 0) -> "Object":
         else:
             l.items.append(parse(tokens, 2))
             while not isinstance(tokens.pop(0), RightBracket):
+                if isinstance(l.items[-1], Spread):
```
Comment (tekknolagi) [inline scrapscript.py:387]: I feel like this should go at the end of the loop

File: scrapscript.py [hunk 9]
```diff
@@ -395,6 +397,8 @@ def parse(tokens: typing.List[Token], p: float = 0) -> "Object":
             assign = parse_assign(tokens, 2)
             l.data[assign.name.name] = assign.value
             while not isinstance(tokens.pop(0), RightBrace):
+                if isinstance(assign.value, Spread):
```
Comment (tekknolagi) [inline scrapscript.py:400]: This too

**Review by tekknolagi — APPROVED** [review]

File: scrapscript.py (hunk 8 above)
Comment (tekknolagi) [reply scrapscript.py:387]: Like after

Comment (tekknolagi) [reply scrapscript.py:387]: jk u right

File: scrapscript.py (hunk 9 above)
Comment (tekknolagi) [reply scrapscript.py:400]: jk u righ

**Comment by tekknolagi** [comment]
looks GREAT to me

**Pull request merged**